evals/record.py

```diff
@@ -341,6 +342,62 @@ def record_final_report(self, final_report: Any):
         logging.info(f"Final report: {final_report}. Logged to {self.event_file_path}")
 
 
+class HttpRecorder(RecorderBase):
+    def __init__(self, url: str, run_spec: RunSpec, batch_size: int = 100):
+        super().__init__(run_spec)
+        self.url = url
+        self.batch_size = batch_size
+        logger.info(f"HttpRecorder initialized with URL {self.url}")
+
+    def _flush_events_internal(self, events_to_write: Sequence[Event]):
+        batch_size = self.batch_size
+        for i in range(0, len(events_to_write), batch_size):
+            batch = list(events_to_write[i : i + batch_size])
+            self._send_event(batch)
+
+    def _send_event(self, events: List[Event]):
+        # Convert the events to dictionaries
+        events_dict = [dataclasses.asdict(event) for event in events]
+
+        logger.debug(f"Sending events: {events_dict}")
+
+        try:
+            # Send the events to the specified URL
+            response = requests.post(self.url, json=events_dict)
+
+            # If the request succeeded, return
+            if response.ok:
+                logger.debug(f"Events sent successfully")
+            else:
+                # If the request failed, log a warning
+                logger.warning(f"Failed to send events: {response.text}")
```

Comment:
We should fail the eval here instead of dropping events

Reply:
Hello @jwang47!

Thank you again for your review. I've taken your feedback into account and made some changes. Here's a brief overview followed by a detailed explanation.

## Quick Summary

I understand your concerns regarding data completeness. However, I believe that completely failing the evaluation due to a small number of event transmission failures may not be the most efficient approach. In the most recent changes, I've made it so that the evaluation can continue even when these failures occur. These failures are also logged for subsequent analysis. In addition, I've introduced a failure threshold. If the proportion of failed events exceeds this threshold, we will then fallback to the Local Recorder. This ensures all events are still recorded and the user is notified of this fallback.

## Detailed Response

Considering the scale of many evaluations (e.g., 10,000 events), I believe the approach I've taken is pragmatic. 

For instance, if only a minimal number of events fail to be sent (like 3 or 7 out of 10,000), the impact on the overall statistical significance of our results is likely to be very small. However, the cost of using GPT-4 (300 tokens ~ per stored message) could be fairly high (around 300 USD).

If we chose to fail the entire evaluation due to these few cases, we would risk incurring significant costs, both financially (especially with paid models like GPT-4) and in terms of time, to re-run the evaluation.

Instead, my approach is to continue the evaluation, log these failed events, and later analyze them to understand and address the underlying issues causing these failures. This way, we can ensure continuity in our evaluations while also maintaining a record of transmission failures for future improvements.

## Demo

### Success

| Screenshot |
|:----------:|
|<img width="593" alt="Screenshot 2023-07-29 at 1 01 25 PM" src="https://github.com/openai/evals/assets/27363885/5561749d-94c6-4247-a6d4-cc22b87dae19">|

### Failure 

| Screenshot 1 | Screenshot 2 |
|:------------:|:------------:|
|<img width="1041" alt="Screenshot 2023-07-29 at 1 02 18 PM" src="https://github.com/openai/evals/assets/27363885/a095366f-a2a6-42f6-afac-d1b43fbd81f8">|<img width="1068" alt="Screenshot 2023-07-29 at 1 02 31 PM" src="https://github.com/openai/evals/assets/27363885/df010875-7f4c-4961-b868-4dcce223c5fa">|

I agree with your concern about not allowing too many events to be dropped. To address this, I've introduced a failure threshold. If the proportion of failed events exceeds this threshold, we will then fallback to the Local Recorder, ensuring all events are still recorded, and notify the user of this fallback. This could strike a sensible balance, allowing us to handle transient issues gracefully while still providing a mechanism to maintain data integrity in the face of more systemic issues.

I'm eager to hear your thoughts on these changes.

Reply:
That makes sense, thanks for the explanation.